[PG/nccl] improvements to eager init

pavanbalaji opened the pull request:
Summary:
In eager-init mode, all p2p communication would use the same collective communicator -- no special p2p communicators are lazily created.  This means that any p2p operations that are issued on that PG would be serialized, for example, with other p2p operations.  This can be surprising to some end users.

This PR throws a warning in such cases, informing users that they should either use batch-mode p2p operations or different PGs for p2p operations in such cases.

Test Plan: CI

Differential Revision: D75157605




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

kwen2501 reviewed (approved):
Risky yet meaningful. Thank you for pushing us to take this first step.

pavanbalaji commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1080,6 +1080,7 @@ void ProcessGroupNCCL::eagerConnectSingleDevice(at::Device device) {
   LOG(INFO) << logPrefix() << "Eagerly connecting nccl backend with device "
             << device;
   initNCCLComm(key, device, OpType::ALLREDUCE);
+  eagerInit = true;
```
All eager-mode PG creations seem to be going through that function, so centralizing it there should be OK, I think.

pavanbalaji commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -3809,23 +3810,77 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::pointToPoint(
+    /* In eagerInit mode, reuse the parent comm.  Do not lazily create
+     * p2p communicators. */
+    bool showSerializationWarning =
+        getCvarBool(TORCH_NCCL_SHOW_EAGER_INIT_P2P_SERIALIZATION_WARNING, true);
```
Moved this to the PG constructor to minimize overhead.

pavanbalaji commented:
> nit: do you mind refine the PR title and description? Linking existing issues, etc? It is a big PR that is worth easier search in future. Thanks!

Done.

msaroufim commented:
/easycla

malfet commented:
/easycla

malfet commented:
/easycla

ngimel commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -3809,23 +3812,75 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::pointToPoint(
+    ncclComm = getNCCLComm(key);
+
+    if (ncclComm == nullptr) {
+      C10_THROW_ERROR(
```
Should be `TORCH_INTERNAL_ASSERT` that automatically adds "Report bug to pytorch"

pavanbalaji committed:
[PG/nccl] improvements to eager init (#154132)

Summary:

Cleanup eager init management, to detect and throw a warning when
multiple p2p are issued on the same PG in eager init mode.

Test Plan: CI

Reviewed By: kwen2501

Differential Revision: D75157605
